mvcc/watcher_group.go

```diff
@@ -78,6 +78,10 @@ func (wb watcherBatch) add(w *watcher, ev mvccpb.Event) {
 // newWatcherBatch maps watchers to their matched events. It enables quick
 // events look up by watcher.
 func newWatcherBatch(wg *watcherGroup, evs []mvccpb.Event) watcherBatch {
+	// Don't allocate anything if we don't have any watchers.
+	if wg.size() == 0 {
```

Comment:
Oh, right, missed that PR. Actually, this particular change is enough to shave off allocations in `TxnEnd` without watches. Maybe that PR was not very convincing without benchmarks.